ts/client/scripts/rebalancer.ts

```diff
@@ -0,0 +1,197 @@
+    userProvider,
+    CLUSTER,
+    MANGO_V4_ID[CLUSTER],
+    {
+      idsSource: 'get-program-accounts',
+    },
+  );
+
+  // Load mango account
+  let mangoAccount = await client.getMangoAccount(
+    new PublicKey(MANGO_ACCOUNT_PK),
+    true,
+  );
+  console.log(
+    `MangoAccount ${mangoAccount.publicKey} for user ${user.publicKey} ${
+      mangoAccount.isDelegate(client) ? 'via delegate ' + user.publicKey : ''
+    }`,
+  );
+  await mangoAccount.reload(client);
+
+  // Load group
+  const group = await client.getGroup(mangoAccount.group);
+  await group.reloadAll(client);
+  const usdcBank = group.getFirstBankByMint(
+    new PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
+  );
+
+  // Loop indefinitely
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    await group.reloadAll(client);
+    mangoAccount = await mangoAccount.reload(client);
+    // console.log(mangoAccount.toString(group, true));
+
+    for (const tp of mangoAccount
+      .tokensActive()
+      .filter((tp) => tp.tokenIndex !== usdcBank.tokenIndex)) {
+      const baseBank = group.getFirstBankByTokenIndex(tp.tokenIndex);
+      const tokenBalance = tp.balanceUi(baseBank);
+
+      const serum3Markets = Array.from(
+        group.serum3MarketsMapByMarketIndex.values(),
+      )
+        // Find correct $TOKEN/$USDC market
+        .filter(
+          (serum3Market) =>
+            serum3Market.baseTokenIndex === tp.tokenIndex &&
+            serum3Market.quoteTokenIndex === usdcBank.tokenIndex,
+        );
+      if (!serum3Markets) {
+        continue;
+      }
+      const serum3Market = serum3Markets[0];
+      const serum3MarketExternal = group.serum3ExternalMarketsMap.get(
+        serum3Market.serumMarketExternal.toBase58(),
+      )!;
+      const maxBaseQuantity = serum3MarketExternal.baseSizeNumberToLots(
+        Math.abs(tokenBalance),
+      );
+      // Skip if quantity is too small
+      if (maxBaseQuantity.eq(new BN(0))) {
+        // console.log(
+        //   ` - Not rebalancing ${tokenBalance} $${baseBank.name}, quantity too small`,
+        // );
+        continue;
+      }
+      console.log(`- Rebalancing ${tokenBalance} $${baseBank.name}`);
+
+      // if balance is negative we want to bid at a higher price
+      // if balance is positive we want to ask at a lower price
+      const price =
+        baseBank.uiPrice *
+        (1 + (tokenBalance > 0 ? -1 : 1) * baseBank.liquidationFee.toNumber());
+      try {
+        const sig = await sendTransaction(
+          client.program.provider as AnchorProvider,
+          [
+            ...(await client.serum3PlaceOrderIx(
+              group,
+              mangoAccount,
+              serum3Market.serumMarketExternal,
+              tokenBalance > 0 ? Serum3Side.ask : Serum3Side.bid,
+              price,
+              Math.abs(tokenBalance),
+              Serum3SelfTradeBehavior.decrementTake,
+              Serum3OrderType.limit,
```

Comment:
why use limit + cancel all instead of ImmediateOrCancel?

Reply:
this is what max suggested

> better to use limit order instead of IOC, you have a higher chance of finding liquidity if you stay on the book for a few slots (arb bots)

> price should be oracle - liquidation fee. you're trying to catch a falling knife, looking at the orderbook will add latency, and the price will have moved most likely by the time you send your order

Reply:
wdyt?

Reply:
Yeah, makes sense. But if we cancel immediately in the same tx we're not resting the limit order on the book for a few slots. The current limit + cancel all in one tx should be effectively a costly immediateorcancel?

But we'd need to change this a good bit to deal with long-lived orders, so that sounds like something for a future PR

Reply:
agreed, I switched back to a immediateorcancel for now